crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -536,6 +536,24 @@ impl SlidingSync {
         let sync_span = Span::current();
 
         stream! {
+            // Drain the internal channel, in case some messages were still present in it.
+            // It's unlikely to happen in a closed Rust world, but it can happen with async
+            // operations over FFI. Let's make sure the internal channel is properly drained
+            // before starting the sync-loop.
+            {
+                let mut internal_channel_receiver_lock = self.inner.internal_channel.1.write().await;
+
+                // *Gurgle, gurgle, gurgle*
+                while let Ok(_) = internal_channel_receiver_lock.try_recv() {}
+
+                // At this point, the receiver must be empty.
+                assert_eq!(
```

Comment:
Also then; the `SyncLoopStop` should remove the internal stream after hitting, sort of like the "close" message from golang channels, which makes it tidy and assures that; if the stream exist, it is either running, or in the process of stopping, not "stopped" or "maybe running".

Reply:
Yeah, but… when the channel is created, the sender is cloned and shared amongst the `SlidingSyncList`, and the receiver is kept by `SlidingSync` itself.
I wanted this to be “immutable” to avoid having a `SlidingSyncList` owning a sender of a receiver that has been dropped for example.